packages/block-editor/src/components/rich-text/index.native.js

```diff
@@ -524,6 +540,43 @@ export class RichText extends Component {
 		return value;
 	}
 
+	componentWillReceiveProps( nextProps ) {
+		this.moveCaretToTheEndIfNeeded( nextProps );
+	}
+
+	moveCaretToTheEndIfNeeded( nextProps ) {
+		const nextRecord = this.formatToValue( nextProps.value );
+		const nextTextContent = getTextContent( nextRecord );
+
+		if ( this.isTouched || ! nextProps.isSelected ) {
+			this.savedContent = nextTextContent;
+			return;
+		}
+
+		if ( nextTextContent === '' && this.savedContent === '' ) {
+			return;
+		}
+
+		// This logic will handle the selection when two blocks are merged or when block is split
+		// into two blocks
+		if ( nextTextContent.startsWith( this.savedContent ) ) {
+			let length = this.savedContent.length;
+			if ( length === 0 && nextTextContent !== this.props.value ) {
+				length = this.props.value.length;
+			}
+
+			this.forceSelectionUpdate( length, length );
+			this.savedContent = nextTextContent;
+		}
+	}
+
+	forceSelectionUpdate( start, end ) {
+		if ( ! this.needsSelectionUpdate ) {
```

Comment:
For this method, I'm confused why we don't call `setState` when `needsSelectionUpdate == true`. The name `forceSelectionUpdate`, to me, implies that it will _always_ update the selection.

Reply:
Hey @mkevins, so it all about resetting it to false, when component renders as I think that it will trigger render one more extra time.

Reply:
I see. I guess the name just threw me off, as the conditional seemed contradictory. As I now understand, it seems this "locks in" the first caller in a race condition.